crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl PaginableRoom for Room {
+    async fn event_with_context(
+        &self,
+        event_id: &EventId,
+        lazy_load_members: bool,
+    ) -> Result<EventWithContextResponse, PaginatorError> {
+        let response = match self.event_with_context(event_id, lazy_load_members, uint!(20)).await {
+            Ok(result) => result,
+
+            Err(err) => {
+                // If the error was a 404, then the event wasn't found on the server; special
+                // case this to make it easy to react to such an error.
+                if let Some(error) = err.as_client_api_error() {
+                    if error.status_code == 404 {
+                        // Event not found
+                        return Err(PaginatorError::EventNotFound(event_id.to_owned()));
+                    }
+                }
+
+                // Otherwise, just return a wrapped error.
+                return Err(PaginatorError::SdkError(err));
+            }
+        };
+
+        Ok(response)
+    }
+
+    async fn messages(&self, opts: MessagesOptions) -> Result<Messages, PaginatorError> {
+        self.messages(opts).await.map_err(PaginatorError::SdkError)
+    }
+}
+
+#[cfg(all(not(target_arch = "wasm32"), test))]
+mod tests {
+    use std::sync::Arc;
+
+    use assert_matches2::assert_let;
+    use async_trait::async_trait;
+    use futures_core::Future;
+    use futures_util::FutureExt as _;
+    use matrix_sdk_test::async_test;
+    use once_cell::sync::Lazy;
+    use ruma::{event_id, room_id, user_id, RoomId, UserId};
+    use tokio::{spawn, sync::Notify};
+
+    use super::*;
+    use crate::test_utils::{assert_event_matches_msg, events::EventFactory};
+
+    #[derive(Default, Clone)]
+    struct DummyRoom {
+        room_ready: Arc<Notify>,
+    }
+
+    impl DummyRoom {
+        /// Unblocks the next request.
+        fn mark_ready(&self) {
+            self.room_ready.notify_one();
+        }
+    }
+
+    static ROOM_ID: Lazy<&RoomId> = Lazy::new(|| room_id!("!dune:herbert.org"));
+    static USER_ID: Lazy<&UserId> = Lazy::new(|| user_id!("@paul:atreid.es"));
+
+    #[async_trait]
+    impl PaginableRoom for DummyRoom {
+        async fn event_with_context(
+            &self,
+            event_id: &EventId,
+            _lazy_load_members: bool,
+        ) -> Result<EventWithContextResponse, PaginatorError> {
+            // Wait for the room to be marked as ready first.
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let event = event_factory.text_msg("hello!").event_id(event_id).into_raw_timeline();
+
+            let before = (0..10)
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
```

Comment:
Personal taste, but I would name these e.g. "before-{i}" or similar and the same for after.